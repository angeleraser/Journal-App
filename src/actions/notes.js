import Swal from "sweetalert2";
import { dataBase } from "../firebase/firebase-config";
import { uploadFile } from "../helpers/fileUpload";
import { fetchUserInfoFromDatabase } from "../helpers/loadNotes";
import { TYPES } from "../types/types";

export const addNewNote = () => {
  return async (dispatch, getState) => {
    // getState returns the current main state and it can be used in async operations
    const {
      auth: { uID },
    } = getState();
    const newEntry = {
      body: "",
      title: "",
      date: new Date().getTime(),
      imgURL: null,
    };
    /*
     Collection journal by user ID 
    */
    const DOCUMENT_REF = await dataBase
      .collection(`${uID}/journal/notes`)
      .add(newEntry);
    dispatch(selectNote(DOCUMENT_REF.id, newEntry));
    dispatch(updateNoteList(DOCUMENT_REF.id, newEntry));
  };
};

export const selectNote = (id, note) => {
  return {
    type: TYPES.notesSelect,
    payload: {
      id,
      ...note,
    },
  };
};

export const setNotes = (notes) => {
  return {
    type: TYPES.notesLoad,
    payload: notes,
  };
};

export const startLoadingNotes = (uID) => {
  return async (dispatch) => {
    const notes = await fetchUserInfoFromDatabase(uID, "journal/notes");
    dispatch(setNotes(notes));
  };
};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const {
      auth: { uID },
    } = getState();
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;
    await dataBase
      .doc(`${uID}/journal/notes/${note.id}`)
      .update(noteToFirestore);
    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: TYPES.notesUpdate,
  payload: {
    id,
    note,
  },
});

export const startUploadingImage = (file) => {
  return async (dispatch, getState) => {
    const { selected } = getState().notes;
    Swal.fire({
      title: "Uploading...",
      text: "Please wait a few seconds...",
      allowOutsideClick: false,
      onBeforeOpen() {
        Swal.showLoading();
      },
    });
    const fileURL = await uploadFile(file);
    selected.imgURL = fileURL;
    Swal.close();
    dispatch(startSaveNote(selected));
  };
};

export const startDeletingNote = (id) => {
  return async (dispatch, getState) => {
    const uID = getState().auth.uID;
    dataBase.doc(`${uID}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: TYPES.notesDelete,
  payload: {
    id,
  },
});

export const cleanNotesAfterLogout = () => ({
  type: TYPES.notesLogoutCleaning,
});

export const updateNoteList = (id, note) => {
  return {
    type: TYPES.notesAddNew,
    payload: {
      id,
      ...note,
    },
  };
};
