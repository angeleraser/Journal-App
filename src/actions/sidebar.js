import { TYPES } from "../types/types";

export const openSidebar = () => ({
  type: TYPES.sidebarOpen,
});

export const closedSidebar = () => ({
  type: TYPES.sidebarClosed,
});
