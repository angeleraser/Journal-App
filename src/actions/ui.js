import { TYPES } from "../types/types";

export const setUiError = (errorMsg) => {
  return {
    type: TYPES.uiSetError,
    payload: errorMsg,
  };
};

export const removeUiError = () => {
  return {
    type: TYPES.uiRemoveError,
  };
};
