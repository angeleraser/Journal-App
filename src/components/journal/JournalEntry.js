import React from "react";

export const JournalEntry = () => {
  return (
    <div className="journal__entry">
      <div
        style={{
          backgroundImage:
            "url(https://images.unsplash.com/photo-1601295863283-f4f2429cdfb3?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=335&q=80)",
          backgroundSize: "cover",
        }}
        className="journal__entry-picture"></div>
      <div className="journal__entry-body">
        <h2>Title</h2>
        <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
      </div>
      <div className="journal__entry-date-box">
        <p>Monday</p>
        <h2>28</h2>
      </div>
    </div>
  );
};
