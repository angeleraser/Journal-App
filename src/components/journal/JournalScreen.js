import React from "react";
import { Main } from "./Main"
import { Sidebar } from "./Sidebar";

export const JournalScreen = () => {
  return (
    <div className="journal__main-content">
      <Sidebar />
      <Main />
    </div>
  );
};
