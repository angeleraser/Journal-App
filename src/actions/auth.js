import Swal from "sweetalert2";
import { FIREBASE, googleAuthProvider } from "..";
import { TYPES } from "../types/types";
import { uiStartLoading, uiFinishLoading } from "./ui";
// NOTA: Si la tarea es asincrona la funcion debe retonar otra funcion que realiza el dispatch, sino un objeto con el tipo de accion y el payload normal.

export const login = (uID, displayName) => ({
  type: TYPES.login,
  payload: {
    uID,
    displayName,
  },
});

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithEmailAndPassword(
          email,
          password
        );
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      Swal.fire("Error", error.message, "error");
      console.log(error);
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().createUserWithEmailAndPassword(
          email,
          password
        );
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startGoogleLogin = () => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithPopup(
          googleAuthProvider
        );
        dispatch(login(user.id, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await FIREBASE.auth().signOut();
    dispatch(logout());
  };
};

export const logout = () => ({
  type: TYPES.logout,
});
