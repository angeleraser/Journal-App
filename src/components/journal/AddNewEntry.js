import React from "react";
import { useDispatch } from "react-redux";
import { addNewNote } from "../../actions/notes";
import { closedSidebar } from "../../actions/sidebar";
import { addIconSvg } from "../../icons/svg";

export const AddNewEntry = () => {
  const dispatch = useDispatch();
  return (
    <button
      onClick={() => {
        dispatch(addNewNote());
        dispatch(closedSidebar());
      }}
      className="journal__new-entry">
      Add New entry
      {addIconSvg}
    </button>
  );
};
