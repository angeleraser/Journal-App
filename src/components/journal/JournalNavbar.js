import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startLogout } from "../../actions/auth";
import { crossIcon } from "../../svg-icons";
import { SidebarButton } from "./SidebarButton";

export const JournalNavbar = () => {
  const dispatch = useDispatch();
  const { name: username, photoURL } = useSelector(({ auth }) => auth);
  const handleLogout = () => {
    dispatch(startLogout());
  };
  return (
    <div className={`journal__sidebar-navbar relative`}>
      <SidebarButton icon={crossIcon} />
      <div className="flex items-center user-profile">
        <span>{username}</span>
        {photoURL && <img src={photoURL} alt="user profile" />}
      </div>
      <button onClick={handleLogout} className="btn logout-button">
        Logout
      </button>
    </div>
  );
};
