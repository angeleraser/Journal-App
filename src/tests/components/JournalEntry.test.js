import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { JournalEntry } from "../../components/journal/JournalEntry";
import { selectNote } from "../../actions/notes";
import { closedSidebar } from "../../actions/sidebar";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Hay que definir el estado inicial igual al de la aplicacion en ese punto
const initState = {
  auth: {
    name: "angeleraser",
    uID: "k1j2g5bk75j23",
    photoURL: null,
  },
  notes: {
    notes: [],
    selected: {
      id: 1234,
      title: "Mi nota",
      body: "body",
      imgURL: "https://sadasdas.com",
    },
  },
  ui: {
    loading: false,
    errorMsg: null,
  },
  sidebar: {
    isOpen: false,
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn();
jest.mock("../../actions/notes", () => {
  return {
    selectNote: jest.fn(),
  };
});
jest.mock("../../actions/sidebar", () => {
  return {
    closedSidebar: jest.fn(),
  };
});
const note = {
  id: "1l6njkl2b4kh124",
  title: "Mi nota",
  body: "Mi cuerpo",
  date: 1241204912841209,
};
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <JournalEntry data={note} />
    </MemoryRouter>
  </Provider>
);

describe("Testing <Sidebar/>", () => {
  // 1-
  test("should match with snap", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should active note and close sidebar", () => {
    wrapper.simulate("click");
    expect(selectNote).toHaveBeenCalled();
    expect(closedSidebar).toHaveBeenCalled();
  });
});
