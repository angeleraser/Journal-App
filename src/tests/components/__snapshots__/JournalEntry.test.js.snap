// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Sidebar/> should match with snap 1`] = `
<div
  className="journal__entry"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://icon-library.com/images/no-image-icon/no-image-icon-0.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <h2>
      Mi nota
    </h2>
    <p>
      Mi cuerpo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <p
      className="mr-1"
    >
      Tuesday
    </p>
    <h2>
      21
    </h2>
  </div>
</div>
`;
